### DISCUSSION TITLE
Inability to Run Microsoft.CodeAnalysis.CSharp.CodeStyle with Unity

### DISCUSSION THREAD

**Discussion by Azirel** [dotnet/roslyn#67861] [Ideas] [answered]
When trying to use Microsoft.CodeAnalysis.CSharp.CodeStyle with Unity, I have encountered an issue. I have ensured that EnforceCodeStyleInBuild is enabled for our project, as Unity runs analyzers on builds. However, adding EnforceCodeStyleInBuild to the .csproj file has had no effect, possibly due to Unity handling .csproj files differently. I am considering reassembling Microsoft.CodeAnalysis.CSharp.CodeStyle with a different name, but I am unsure if this is a viable solution given the complexity of the Roslyn project. Can anyone provide insights on whether this approach is worth trying and if it can be done within a reasonable timeframe?

**Comment by jasonmalinowski** [comment]
What would you be attempting to gain by renaming it?

**Comment by Azirel** [comment]
@jasonmalinowski I attempt to bypass EnforceCodeStyleInBuild condition since as far as I'm aware it is hadrcoded into compiler and checks for specific namespace. Meaning if analyzer id is Microsoft.CodeAnalysis.CSharp.CodeStyle then EnforceCodeStyleInBuild would be applied to it by compiler itself.

**Reply by mavasani** [reply]
This is not correct. This MSBuild property adds an explicit analyzer reference to the CodeStyle analyzer assembly so it is passed as a command line argument to the compiler. There is no special casing in the compiler based on namespace.

**Reply by Azirel** [reply]
Does this mean that if I pass Microsoft.CodeAnalysis.CSharp.CodeStyle analyzer manually as cli argument it will generate diagnostic report? In this case it has to work already, but it doesn't.

**Comment by Azirel** [comment]
In fact, I would've test with creating mock analyzer, name it Microsoft.CodeAnalysis.CSharp.CodeStyle and look at its behaviour first.
And that's what I'm about to do now. I'll post an update on it here.

**Comment by Azirel** [comment]
Ok. It seems like renaming it has no effect. I made new custom analyzer with Microsoft.CodeAnalysis.CSharp.CodeStyle name and I get errors on build.

**Comment by Azirel** [comment]
Any other ideas on how to make Microsoft.CodeAnalysis.CSharp.CodeStyle work with Unity?

**Comment by mavasani** [comment]
@Azirel did you follow the steps mentioned here: https://learn.microsoft.com/en-us/dotnet/fundamentals/code-analysis/overview?tabs=net-7#enable-on-build

**Reply by Azirel** [reply]
I'm aware of these instructions. The problem is I work with Unity hence those instructions kinda irrelevant.
Unity does not work with .editorconfig file at all. As name suggests it's for editor, not for build. Unity works with .ruleset file for setting rules severity. And I did set EnforceCodeStyleInBuild property to true in my .csproj file already.

**Reply by mavasani** [reply]
I am still confused about the actual issue. Are you setting the required IDExxxx rule IDs to warnings in the ruleset file? If so, are you not seeing these IDExxxx warnings in Unity editor while editing code as well as command line build OR you see them on command line build but not in the editor.

**Reply by Azirel** [reply]
Are you setting the required IDExxxx rule IDs to warnings in the ruleset file? - yes
Are you not seeing these IDExxxx warnings in Unity editor while editing code -yes

Little context: each time u make a change in code - Unity makes a build and produces all errors/warnings inside it's editor. So if u have any errors/warnings u a going to see them in Unity editor. I can setup roslyn analyzers to work in Unity, except Microsoft.CodeAnalysis.CSharp.CodeStyle.

**Comment by Azirel** [comment, marked as answer]
Problem solved. Issue was in .dll version. It has to be in sync with Microsoft.CodeAnalisys.dll that unity uses under the hood.
